### ISSUE TITLE
Temperature misreported when T1 active, dual extruders, Marlin 2.0.9.2 based firmware on Sovol SV04

### ISSUE THREAD

**Issue by tringenbach** [OctoPrint/OctoPrint#4890]
### The problem

For my printer and it's latest vender firmware, when the second extruder is active, the temperature of the first extruder (T0) is misreported in Octoprint as the temperature of the second extruder (T1). This seems to be because Marlin's status line is also very confusing (and also wrong?).

I have a Sovol SV04 IDEX printer, that is running Sovol's modified firmware based off of Marlin 2.0.9.2. I've looked through their modifications, and maybe I missed something, but I don't think they come into play for this bug. I haven't checked if it happens in other versions of Marlin.

The temperature autoreport comes back like this:
```
 Recv:  T0:220.36 /220.00 B:60.03 /60.00 T0:50.44 /0.00 T1:220.36 /220.00 @:57 B@:37 @0:0 @1:57
```

Notice that `T0` is listed twice!

The first `T0` seems to actually be the active extruder. So this issue only happens after issuing a `T1` command. At that point, the first `T0` is actually `T1`.

So to report the temperature correctly when T1 is active, Octoprint needs to ignore the first `T0` (which is really `T1`) and use the second one. Unfortunately for me, there's a check in the parser to avoid overriding a sensor once its value is seen.

I only recently started using Octoprint, so I haven't tried older versions, but I see that in 2ecc1c944f9f00bbd4a3e273f0c591e74d06305d this was added:

```diff
diff --git a/src/octoprint/util/comm.py b/src/octoprint/util/comm.py
index 9677002ae..d06b9529b 100644
--- a/src/octoprint/util/comm.py
+++ b/src/octoprint/util/comm.py
@@ -6268,6 +6268,10 @@ def parse_temperature_line(line, current):
     for match in re.finditer(regex_temp, line):
         values = match.groupdict()
         sensor = values["sensor"]
+        if sensor in result:
+            # sensor already seen, let's not overwrite stuff
+            continue
+
```

I assume that was added for a good reason and I shouldn't just submit a PR removing it; that would break something else.

Perhaps the correct fix is to special case the first `T0` (or the second `T0`) and let it overwrite, but not anything else. But I don't have the breath of 3d printer experience necessary to say for sure.

### Did the issue persist even in safe mode?

Yes, it did persist

### Version of OctoPrint

1.9.2

### Operating system running OctoPrint

OctoPi  Build 2023.07.20.144556, based on OctoPi 1.0.0, running on Raspberry Pi 4 Model B Rev 1.5

### Printer model & used firmware incl. version

Sovol SV04, Sovol's version of Marlin 2.0.9.2 

### Browser and version of browser, operating system running browser

Chrome 116, MacOS

### Checklist of files to include below

- [X] Systeminfo Bundle (always include!)
- [ ] Contents of the JavaScript browser console (always include in cases of issues with the user interface)
- [ ] Screenshots and/or videos showing the problem (always include in case of issues with the user interface)
- [ ] GCODE file with which to reproduce (always include in case of issues with GCODE analysis or printing behaviour)

### Additional information & file uploads

[octoprint-systeminfo-20231001095603.zip](https://github.com/OctoPrint/OctoPrint/files/12777416/octoprint-systeminfo-20231001095603.zip)

```
Recv:  T0:26.22 /0.00 B:42.48 /0.00 T0:26.22 /0.00 T1:34.71 /0.00 @:0 B@:0 @0:0 @1:0
[...]
Recv: echo:Active Extruder: 1
Recv: ok
[...]
Recv:  T0:34.61 /0.00 B:42.46 /0.00 T0:26.14 /0.00 T1:34.61 /0.00 @:0 B@:0 @0:0 @1:0
```

<img width="997" alt="image" src="https://github.com/OctoPrint/OctoPrint/assets/381141/faceb0e2-e4ae-48b2-9045-51389b1fa34b">

**Comment by foosel** [comment]
Hey there...

first of all thank you for that thorough report and analysis.

You are right that this looks indeed like the firmware is reporting T1 as T0 here after switching the active extruder. That however is something I'd consider a firmware bug. The whole temperature parsing is a horrible mess since a ton of firmware variants out there do their reports ever so slightly differently, however, so far an agreed upon behaviour seems to have been that in case of multiple extruders, T0 is always the first, T1 is always the second and so on. If no T0 is provided but rather a T, this is actually T0 (Smoothieware) and if T and T0, T1, etc are present, T is the temperature of the currently selected extruder.

See also the canonicalization logic that's been extracted from firmwares that OctoPrint has had to interface over the past decade:

https://github.com/OctoPrint/OctoPrint/blob/ed4a2646fb4e2904892c895580192987242834c8/src/octoprint/util/comm.py#L6265-L6351

Multiple entries for one key (two T0 as seen here) is something that's completely new and frankly something I'd like to declare a firmware bug. T and T0 would be something that has been already established as behaviour in the wild and is something that OctoPrint would handle correctly here, but T0 and T0 is just plain wrong in my opinion, and I'm not going to implement support for that (figuring out which one is the actual T0 feels like guesswork anyhow tbh). 

In general we've seen firmware reporting weird stuff at the end of temperature lines that could be repetitions of existing entries, and which have proven to be non-sense with regards to the function of the temperature report, so those lines you mentioned up there are in place to just ignore that stuff and keep it from messing with temperature visualization on affected firmware.

Long story short, this is a firmware issue and something I'd rather not add yet another exception for in the temperature canonicalization code, which is also way more complicated than it has any right to be, but had to be thanks to the mess that is 3d printer firmwares out there. It would however be possible to work around this with a simple plugin utilizing [the `octoprint.comm.protocol.gcode.received` hook](https://docs.octoprint.org/en/master/plugins/hooks.html#octoprint-comm-protocol-gcode-received) until the firware in question is fixed.
